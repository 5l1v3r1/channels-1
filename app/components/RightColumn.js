// @flow
import React, { Component } from 'react';
import { IconChannels } from './core/Icons';
import TitleBar from './TitleBar';
import Player from './Player';
import RatingBar from './RatingBar';
import Comments from '../containers/Comments';
import RatingType from '../dataTypes/ratingType';
import styles from './RightColumn.scss';

export default class RightColumn extends Component {
  props: {
    activeVideo: any,
    rating: ?RatingType,
    fetchRatings: () => Promise<any>,
    postRating: (type: string) => Promise<any>,
    updateActiveVideoCounts: (prevState: RatingType, rating: string) => void
  };

  state: {
    isPosting: boolean
  };

  postRating: (rating: string) => void

  constructor(props: any) {
    super(props);
    this.state = { isPosting: false };
    this.postRating = this.postRating.bind(this);
  }

  componentWillReceiveProps(nextProps: any) {
    const { id } = this.props.activeVideo;
    const { id: nextId } = nextProps.activeVideo;
    if (id !== nextId) {
      this.props.fetchRatings();
    }
  }

  postRating(rating: string): void {
    this.setState({ isPosting: true });
    this.props.postRating(rating)
      .then(
        prevRating => this.props.updateActiveVideoCounts(prevRating, rating),
        () => undefined
      )
      .then(() => this.setState({ isPosting: false }));
  }

  render() {
    const {
      activeVideo,
      rating
    } = this.props;
    const {
      isPosting
    } = this.state;
    if (!Object.keys(activeVideo).length) {
      return (
        <div className={styles.rightColumn}>
          <IconChannels styles={styles} />
        </div>
      );
    }

    return (
      <div className={styles.rightColumn}>
        <TitleBar title={activeVideo.title} />
        <Player id={activeVideo.id} />
        <RatingBar
          likeCount={activeVideo.likeCount}
          dislikeCount={activeVideo.dislikeCount}
          rating={rating}
          isPosting={isPosting}
          postRating={this.postRating}
        />
        <Comments />
      </div>
    );
  }
}
