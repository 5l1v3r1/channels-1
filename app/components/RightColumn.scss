@import '../app.global.scss';

.rightColumn {
  position: relative;
  display: flex;
  flex-direction: column;
  width: 50%;
  height: 100%;
  background-color: $dark-black;
}

.iconChannels {
  position: absolute;
  top: 50%;
  left: 50%;
  transform: translate(-50%, -50%);
  width: 50px;
}

.screen {
  fill: none;
}

.frame {
  fill: $red;
}
