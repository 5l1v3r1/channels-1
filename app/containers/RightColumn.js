import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import * as activeVideoActions from '../actions/activeVideo';
import * as ratingsActions from '../actions/ratings';
import RightColumn from '../components/RightColumn';
import { getActiveVideoRating } from '../reducers/index';

function mapStateToProps(state) {
  return {
    activeVideo: state.activeVideo,
    rating: getActiveVideoRating(state)
  };
}

const actions = {
  ...activeVideoActions,
  ...ratingsActions
};

function mapDispatchToProps(dispatch) {
  return bindActionCreators(actions, dispatch);
}

export default connect(mapStateToProps, mapDispatchToProps)(RightColumn);
