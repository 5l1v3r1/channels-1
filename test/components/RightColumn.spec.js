import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import RightColumn from '../../app/components/RightColumn';

jest.mock('../../app/containers/Comments', () => 'Comments');

function setup() {
  const props = {
    activeVideo: {},
    rating: {},
    fetchRatings: jest.fn(),
    postRating: jest.fn(),
    updateActiveVideoCounts: jest.fn()
  };

  const enzymeWrapper = shallow(<RightColumn {...props} />);

  return {
    props,
    enzymeWrapper
  };
}

describe('RightColumn', () => {
  describe('when no active video', () => {
    it('should render self', () => {
      const { enzymeWrapper } = setup();
      const tree = toJson(enzymeWrapper);

      expect(tree).toMatchSnapshot();
    });
  });

  describe('when active video', () => {
    it('should render self', () => {
      const { enzymeWrapper } = setup();
      enzymeWrapper.setProps({
        activeVideo: {
          id: 'XsFQEUP1MxI',
          title: 'Unit testing in JavaScript Part 2 - Your first tests',
          thumbnail: 'https://i.ytimg.com/hqdefault.jpg',
          description: 'Today, we are continuing our journey on unit testing in JavaScript.',
          publishedAt: '6 days ago',
          viewCount: '10,276',
          likeCount: '624',
          dislikeCount: '3'
        }
      });
      const tree = toJson(enzymeWrapper);

      expect(tree).toMatchSnapshot();
    });
  });

  it('should call fetchRatings on activeVideo id change', () => {
    const { props, enzymeWrapper } = setup();

    expect(props.fetchRatings).toHaveBeenCalledTimes(0);

    enzymeWrapper.setProps({
      activeVideo: {
        id: 'XsFQEUP1MxI'
      }
    });

    expect(props.fetchRatings).toHaveBeenCalledTimes(1);

    enzymeWrapper.setProps({
      activeVideo: {
        id: 'Eu35xM76kKY'
      }
    });

    expect(props.fetchRatings).toHaveBeenCalledTimes(2);

    enzymeWrapper.setProps({
      activeVideo: {
        id: 'Eu35xM76kKY'
      }
    });

    expect(props.fetchRatings).toHaveBeenCalledTimes(2);
  });
});
