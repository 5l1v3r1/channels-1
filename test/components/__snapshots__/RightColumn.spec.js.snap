// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RightColumn when active video should render self 1`] = `
<div
  className="rightColumn"
>
  <TitleBar
    title="Unit testing in JavaScript Part 2 - Your first tests"
  />
  <Player
    id="XsFQEUP1MxI"
  />
  <RatingBar
    dislikeCount="3"
    isPosting={false}
    likeCount="624"
    postRating={[Function]}
    rating={Object {}}
  />
  <Comments />
</div>
`;

exports[`RightColumn when no active video should render self 1`] = `
<div
  className="rightColumn"
>
  <Component
    styles={Object {}}
  />
</div>
`;
